streams/src/main/java/org/apache/kafka/streams/processor/assignment/KafkaStreamsState.java

```
@@ -0,0 +1,97 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.processor.assignment;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.SortedSet;
+import org.apache.kafka.streams.processor.TaskId;
+import org.apache.kafka.streams.state.HostInfo;
+
+/**
+ * A read-only metadata class representing the current state of each KafkaStreams client with at least one StreamThread participating in this rebalance
+ */
+public interface KafkaStreamsState {
+    /**
+     * @return the processId of the application instance running on this KafkaStreams client
+     */
+    ProcessID processId();
+
+    /**
+     * Returns the number of processing threads available to work on tasks for this KafkaStreams client,
+     * which represents its overall capacity for work relative to other KafkaStreams clients.
+     *
+     * @return the number of processing threads on this KafkaStreams client
+     */
+    int numProcessingThreads();
+
+    /**
+     * @return the set of consumer client ids for this KafkaStreams client
+     */
+    SortedSet<String> consumerClientIds();
+
+    /**
+     * @return the set of all active tasks owned by consumers on this KafkaStreams client since the previous rebalance
+     */
+    SortedSet<TaskId> previousActiveTasks();
+
+    /**
+     * @return the set of all standby tasks owned by consumers on this KafkaStreams client since the previous rebalance
+     */
+    SortedSet<TaskId> previousStandbyTasks();
+
+    /**
+     * Returns the total lag across all logged stores in the task. Equal to the end offset sum if this client
+     * did not have any state for this task on disk.
+     *
+     * @return end offset sum - offset sum
+     *                    Task.LATEST_OFFSET if this was previously an active running task on this client
```

I realize I forgot this in the KIP, but we should throw an exception if this method is called and the user did not request task lags be computed. Not sure what exception type makes the most sense here, maybe  IllegalArgumentException or UnsupportedOperationException?

Whichever you choose, just add a `@throws` statement to the javadocs here

---

ditto for the `#prevTasksByLag` and `#statefulTasksToLagSums` methods below

---

IllegalArgumentException usually refers to an argument that was passed in, as opposed to some global misconfiguration, so I'll go for UnsupportedOperationException unless someone else feels strongly about it.

---

Makes sense. We definitely abuse IllegalArgumentException to mean more than that elsewhere in Streams but no reason to perpetuate that bad habit lol